scholly/service/core/push.js

```diff
@@ -0,0 +1,186 @@
+
+/**
+ * This function takes in a user's GitHub username and their PAT and deploys the built React base app to
+ * their GitHub pages repository.
+ *
+ * @param ghUsername the GitHub username of the account the app needs to be deployed to.
+ * @param ghToken the Personal Access Token (PAT) of the user's account with push permissions.
+ * @throws Error if the push failed.
+ */
+async function pushBuiltAppToPages(ghUsername, ghToken) {
+  console.log(ghUsername);
+  console.log(ghToken);
+  await makeGHRepo(ghUsername, ghToken);
+  await ghpages.publish(
+    '../base/build',
+    {
+      branch: 'main',
+      repo: `https://${ghUsername}:${ghToken}@github.com/${ghUsername}/${ghUsername}.github.io.git`,
+      message: 'Deployed with Researchify :)',
+      user: {
+        name: 'Researchify',
+        email: 'researchify@bot.com',
+      },
+    },
+    (err) => {
+      if (err) {
+        winston.error(
+          `Failed to push built app to GitHub Pages for ${ghUsername}`,
+          err
+        );
+        throw err;
+      } else {
+        winston.info(`Successfully deployed app for ${ghUsername}`);
+      }
+      cleanupBuild(); // TODO this is not working for some reason
+    }
+  );
+  await buildPages(ghUsername, ghToken);
+}
+
+/**
+ * Utility function to remove the generated build directory after the React base app has been built.
+ * We won't be requiring it anymore and can thus be removed.
+ */
+function cleanupBuild() {
+  fs.rm(`${PATH_TO_BASE_REACT_APP}/build`, { recursive: true }, (err) => {
+    if (err) {
+      winston.error('Failed to remove built directory.');
+    }
+  });
+}
+
+async function makeGHRepo(ghUsername, ghToken) {
+  // check if the repo exists
+  const repoName = ghUsername + '.github.io';
+  const searchQuery = 'q=user:' + ghUsername + ' ' + repoName + ' in:name';
+  const options = {
+    Authorization: 'token ' + ghToken,
+    Accept: 'application/vnd.github.v3+json',
+  };
+  let repoResponse;
+  try {
+    repoResponse = await axios.get(
+      'https://api.github.com/search/repositories?' + searchQuery,
+      {
+        headers: options,
+      }
+    );
+  } catch (err) {
+    console.log(err);
+  }
+
+  if (repoResponse.data.total_count === 1) {
+    console.log('repo exists already');
+  } else {
+    // if it doesn't, make one
+    const createRepoBody = {
+      name: repoName,
+      private: false, // free accounts can only use a public repo for GH pages
+      auto_init: true,
+    };
+    try {
+      await axios({
+        url: 'https://api.github.com/user/repos',
+        method: 'post',
+        headers: options,
+        data: createRepoBody,
+      });
+      console.log('Repo was created successfully');
+    } catch (err) {
+      console.log(err);
+    }
+  }
+
+  // set up pages for the repo regardless of whether its new or not
+  // see https://docs.github.com/en/rest/reference/repos#pages
+  try {
+    const pagesResponse = await axios.get(
```

Comment:
Why do we need to GET the GitHub webpages?

Reply:
This checks whether the user already has a GH pages website set up, if they do have one set up, we don't have to set it up for them. If they don't have one, this axios call will return an error and we'll set it up for them. I think if the repo name is {user}.github.io it should automatically be set up, but this could be helpful like you mentioned earlier if it becomes a requirement that the repo needs to be called something else.

Reply:
ahh true